### ISSUE TITLE
[AOTInductor] mmap-backed constant mappings survive runner destruction

### ISSUE THREAD

**Issue by sujuyu** [pytorch/pytorch#193915]
### 🐛 Describe the bug

I reproduced this behavior in a rolling AOTInductor model-loading workload. I reviewed the report below and confirmed that the `/proc/self/maps` behavior matches my local testing.

I would like to work on this issue. If the scope looks actionable, could a maintainer please assign it to me? I will submit a regression test and a fix after the issue is marked actionable.

> The following report was drafted with Codex assistance.
>
> ## Summary
>
> When an AOTInductor package is compiled with
> `aot_inductor.force_mmap_weights=True`, loading the package creates a
> file-backed mapping for the constant blob at the end of `.wrapper.so`.
>
> After the loaded runner is deleted and Python garbage collection runs, this
> weight mapping remains in `/proc/self/maps`. Repeated load/delete cycles
> accumulate additional mappings instead of returning to the original
> baseline.
>
> ## Reproduction
>
> ```python
> import gc
> import os
> import pathlib
> import tempfile
> import zipfile
>
> import torch
>
>
> class Model(torch.nn.Module):
>     def __init__(self):
>         super().__init__()
>         self.weight = torch.nn.Parameter(torch.randn(16, 16))
>
>     def forward(self, x):
>         return x @ self.weight
>
>
> inputs = (torch.randn(4, 16),)
> ep = torch.export.export(Model(), inputs)
>
> with tempfile.TemporaryDirectory() as tmp_dir:
>     package_path = os.path.join(tmp_dir, "model.pt2")
>     torch._inductor.aoti_compile_and_package(
>         ep,
>         package_path=package_path,
>         inductor_configs={"aot_inductor.force_mmap_weights": True},
>     )
>
>     with zipfile.ZipFile(package_path) as package:
>         wrappers = [
>             info
>             for info in package.infolist()
>             if info.filename.endswith(".wrapper.so")
>         ]
>
>     assert len(wrappers) == 1
>     wrapper_name = pathlib.PurePosixPath(wrappers[0].filename).name
>     wrapper_size = wrappers[0].file_size
>
>     def count_weight_mappings():
>         count = 0
>         with open("/proc/self/maps") as maps:
>             for line in maps:
>                 if f"/{wrapper_name}" not in line:
>                     continue
>
>                 address_range, permissions, offset, *_ = line.split()
>                 start, end = (
>                     int(value, 16)
>                     for value in address_range.split("-")
>                 )
>
>                 # Select the writable mapping that reaches the file tail,
>                 # where the AOTI constant blob is stored.
>                 if (
>                     permissions == "rw-p"
>                     and int(offset, 16) + end - start >= wrapper_size
>                 ):
>                     count += 1
>         return count
>
>     gc.collect()
>     baseline = count_weight_mappings()
>     print("baseline:", baseline)
>
>     for iteration in range(3):
>         runner = torch._inductor.aoti_load_package(package_path)
>         print("after load:", iteration, count_weight_mappings())
>
>         del runner
>         gc.collect()
>         print("after delete:", iteration, count_weight_mappings())
> ```
>
> ## Actual behavior
>
> Loading creates a new file-tail weight mapping. Deleting the runner does not
> return the mapping count to its baseline, and subsequent load/delete cycles
> accumulate additional mappings.
>
> The check above deliberately excludes the normal ELF load segments retained
> by the dynamic linker. It only counts writable mappings whose file range
> reaches the end of `.wrapper.so`, which identifies the mmap-backed constant
> blob.
>
> ## Expected behavior
>
> Destroying the final owner of an AOTInductor runner should release its
> mmap-backed constant mapping. After `del runner` and `gc.collect()`, the
> mapping count should return to the pre-load baseline.
>
> ## Impact
>
> This is primarily a file-backed virtual-address-space mapping leak, not
> necessarily an equivalent resident-RAM leak because the mapped pages remain
> reclaimable.
>
> It is nevertheless relevant to long-lived inference processes that roll
> large AOTI models on and off. With multi-gigabyte constant blobs, repeated
> model replacement leaves tens of GiB of stale virtual mappings and associated
> VMA/page-table state until the process exits.
>
> ## Environment
>
> - Originally observed with PyTorch 2.12.1+cu129 and 5-8 GiB CUDA AOTI packages
> - Confirmed on Linux with PyTorch main at commit
>   `2472115902865f117c5ecbae859239402b16afbc`
> - Built version: `2.15.0a0+git2472115`
> - The minimal reproduction uses CPU; CUDA is not required

### Versions

Collecting environment information...
PyTorch version: 2.15.0a0+git2472115
Is debug build: False
CUDA used to build PyTorch: 12.9
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 12.0.0-3ubuntu1~20.04.5
CMake version: version 4.4.2
Libc version: glibc-2.31

Python version: 3.11.14 | packaged by conda-forge | (main, Jan 26 2026, 23:48:32) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-146-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.9.41
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe
GPU 4: NVIDIA A100 80GB PCIe
GPU 5: NVIDIA A100 80GB PCIe

Nvidia driver version: 575.57.08
cuDNN version: Probably one of the following:
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-127
Thread(s) per core:              1
Core(s) per socket:              64
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           49
Model name:                      AMD EPYC 7H12 64-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         1410.339
CPU max MHz:                     2600.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        5190.30
Virtualization:                  AMD-V
L1d cache:                       4 MiB
L1i cache:                       4 MiB
L2 cache:                        64 MiB
L3 cache:                        512 MiB
NUMA node0 CPU(s):               0-63
NUMA node1 CPU(s):               64-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate sme ssbd mba sev ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.15.0a0+git2472115
[conda] Could not collect

cc @chauhang @penguinwu @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @ydwu4 @desertfire @yushangdi @jataylo @iupaikov-amd

**Comment by sujuyu** [comment]
Thanks for assigning this to me, @yushangdi! Should this issue also be marked as actionable before I submit the PR?

**Comment by sujuyu** [comment]
> Thanks for assigning this to me, [@yushangdi](https://github.com/yushangdi)! Should this issue also be marked as actionable before I submit the PR?

Hi @desertfire , would you mind taking a quick look at this issue when you have a chance? I have a small fix and regression test ready locally, but I'm waiting for the actionable label before opening the PR per the contribution guidelines. Thanks!

**Comment by desertfire** [comment]
@sujuyu , you can just open an PR. The analysis makes sense to me.
